apps/web/src/hooks/usePermitStatus.ts

```diff
@@ -0,0 +1,296 @@
+import {
+  AllowanceTransfer,
+  Permit,
+  Permit2ABI,
+  generatePermitTypedData,
+  getPermit2Address,
+} from '@pancakeswap/permit2-sdk'
+import { Currency, CurrencyAmount, Token } from '@pancakeswap/swap-sdk-core'
+import { ethereumTokens } from '@pancakeswap/tokens'
+import { Permit2Signature } from '@pancakeswap/universal-router-sdk'
+import { useQuery } from '@tanstack/react-query'
+import { SLOW_INTERVAL } from 'config/constants'
+import { useCallback, useMemo } from 'react'
+import { publicClient } from 'utils/client'
+import { Address, isAddressEqual, zeroAddress } from 'viem'
+import { useSignTypedData } from 'wagmi'
+import { SendTransactionResult } from 'wagmi/actions'
+import useAccountActiveChain from './useAccountActiveChain'
+import { useActiveChainId } from './useActiveChainId'
+import { ApprovalState, useApproveCallback } from './useApproveCallback'
+import useCurrentBlockTimestamp from './useCurrentBlockTimestamp'
+import useTokenAllowance from './useTokenAllowance'
+
+type PermitStatus<T extends Token> = {
+  // allowance to permit2 address
+  allowance?: CurrencyAmount<T>
+  // permit amount through permit2
+  permitAmount?: CurrencyAmount<T>
+  // permit expiration
+  expiration: number
+  nonce: number
+}
+
+// return current wallet's permit status
+export const usePermitStatus = (token?: Token, owner?: Address, spender?: Address) => {
+  const allowance = useTokenAllowanceToPermit2(token, owner)
+  const permit = useTokenPermit(token, owner, spender)
+
+  return useMemo<PermitStatus<Token>>(
+    () => ({
+      allowance,
+      ...permit,
+    }),
+    [allowance, permit],
+  )
+}
+
+const useTokenAllowanceToPermit2 = (token?: Currency, owner?: Address) => {
+  const { chainId } = useActiveChainId()
+  const { allowance } = useTokenAllowance(
+    token?.isNative ? undefined : (token as Token),
+    owner,
+    getPermit2Address(chainId),
+  )
+  return allowance
+}
+
+const useTokenPermit = (token?: Token, owner?: Address, spender?: Address) => {
+  const { chainId } = useActiveChainId()
+
+  const inputs = useMemo<[Address, Address, Address]>(
+    () => [owner ?? zeroAddress, token?.address ?? zeroAddress, spender ?? zeroAddress],
+    [owner, spender, token?.address],
+  )
+  const defaultPermit = useMemo(() => {
+    return {
+      expiration: 0,
+      permitAmount: token ? CurrencyAmount.fromRawAmount(token, '0') : undefined,
+      nonce: 0,
+    }
+  }, [token])
+
+  const { data: permit } = useQuery(
+    ['/token-permit/', chainId, token?.address, owner, spender],
+    () =>
+      chainId && !token?.isNative
+        ? publicClient({ chainId }).readContract({
+            abi: Permit2ABI,
+            address: getPermit2Address(chainId),
+            functionName: 'allowance',
+            args: inputs,
+          })
+        : undefined,
```

Comment:
I recall that `useQuery` will throw warning if the query is returning `undefined`. Probably better to throw error if some of the requirements aren't met